Hash-based challenge parts navigation

loic-brtd opened the pull request:
Choo choo! :steam_locomotive: 

In response to #1319, I'm giving a go at implementing hash-based navigation between challenge parts. This would allow us to reference a specific challenge part by its URL. Example: https://thecodingtrain.com/challenges/40-word-counter#part-3.

A few observations on my current implementation : 
- `/challenges/40-word-counter` takes us to the 1st part of the challenge
- `/challenges/40-word-counter#part-1` takes us to the 1st part of the challenge
- `/challenges/40-word-counter#part-2` takes us to the 2nd part of the challenge
- `/challenges/40-word-counter#part-3` takes us to the 3rd part of the challenge

Whatever hash value is passed to the URL by the user, the URL is not modified (unless we click on another part link or prev/next button of course) : 
- `/challenges/40-word-counter#part-42` takes us to the 3rd (last) part of the challenge (`#part-42` stays in the URL)
- `/challenges/40-word-counter#part-abc` takes us to the 1st part of the challenge (`#part-abc` stays in the URL)

I added this on challenge and track pages containing multi-part challenges.

Previews : 
- https://deploy-preview-1380--codingtrain.netlify.app/challenges/40-word-counter
- https://deploy-preview-1380--codingtrain.netlify.app/tracks/games/72-frogger#part-3

I'm still working on cleaning up some of the code. I welcome any feedback on this feature :grinning:

shiffman commented:
I haven't reviewed the code closely, but I've tested out the implementation in the deploy branch and I think it's working terrifically!  Maybe @fturmel or @dipamsen might have a few minutes to take a look at the code?

loic-brtd commented:
Thanks @shiffman, I made a finishing touch to the code/documentation and marked this PR as ready for review :smiley:

fturmel commented:
Fantastic job @loic-brtd, especially noticing the track stops behavior!

I've noticed two things:

* There's a a regression in the mobile view. When you open the panel to navigate between parts, when selecting one the panel doesn't close anymore.
* I personally might prefer getting rid of `#part-1` since it is equivalent to not having an anchor at all

Good stuff!

loic-brtd commented:
Thanks @fturmel !

> * There's a a regression in the mobile view. When you open the panel to navigate between parts and select one, the panel doesn't close anymore. Same issue in tracks.

Good catch! I'm gonna fix this.

> * I personally might prefer getting rid of `#part-1` since it is equivalent to not having an anchor at all

Why not :) I'd like to have multiple opinions on this

loic-brtd commented:
@fturmel Btw I just noticed that even on the live website (main branch), the timeline doesn't close on mobile in tracks. Seems like I forgot to handle that case when implementing the feature back then (#743) :upside_down_face: . Anyway, I pushed a fix :smiley:

shiffman commented:
I think I would agree that the main link to the challenge should just default to part 1 and a separate #part-1 is not needed! However, I'm fine with both working!

loic-brtd commented:
@fturmel @shiffman I removed `#part-1` in links for the first part of challenges.

I also simplified the logic : if the URL hash value matches `#part-{partNumber}` where `1 <= partNumber <= number of parts` then we're taken to this part, else we're taken to the first part.

fturmel reviewed (approved):
Looks great to me!

shiffman commented:
This is incredible work, thank you for your contribution @loic-brtd! Thrilled to merge this (and thank you for your patience!)
